dimod/binary_quadratic_model/model.py

```diff
@@ -0,0 +1,1099 @@
+                Variables that have an interaction and their quadratic bias. If a dict, the keys
+                should be 2-tuples of the variables and the values should be their corresponding
+                bias. Can also be an iterable of 3-tuples. Each interaction in quadratic should be
+                unique - that is if `(u, v)` is a key in quadratic, then `(v, u)` should not be.
+                The variables can be any python object that could be used as a key in a dict.
+                Many methods and functions expect the biases to be numbers but this is not
+                explicitly checked.
+
+            vartype (:class:`.Vartype`, optional, default=None):
+                The vartype of the given bias. If None will be the same vartype as the binary
+                quadratic model. If given, should be :class:`.Vartype.SPIN` or
+                :class:`.Vartype.BINARY`.
+
+        Examples:
+            >>> bqm = dimod.BinaryQuadraticModel({}, {}, 0.0, dimod.SPIN)
+            >>> bqm.add_interactions_from({('a', 'b'): -.5})
+            >>> bqm.quadratic
+            {('a', 'b'): -.5}
+
+            Variables that already exist have their bias added.
+
+            >>> bqm = dimod.BinaryQuadraticModel({}, {('b', 'a'): -.5}, 0.0, dimod.SPIN)
+            >>> bqm.add_interactions_from({('a', 'b'): -.5})
+            >>> bqm.quadratic
+            {('b', 'a'): -1.}
+
+        """
+        if isinstance(quadratic, dict):
+            for (u, v), bias in iteritems(quadratic):
+                self.add_interaction(u, v, bias, vartype=vartype)
+        else:
+            try:
+                for u, v, bias in quadratic:
+                    self.add_interaction(u, v, bias, vartype=vartype)
+            except TypeError:
+                raise TypeError("expected 'quadratic' to be a dict or an iterable of 3-tuples.")
+
+    def remove_variable(self, v):
+        """Remove the variable v and all of its interactions.
+
+        Args:
+            v (variable):
+                A variable in the binary quadratic model.
+
+        Notes:
+            If the given variable is not in the binary quadratic model, this function does nothing.
+
+        Examples:
+            >>> bqm = dimod.BinaryQuadraticModel({'a': 1., 'b': 2.}, {}, 0.0, dimod.SPIN)
+            >>> bqm.remove_variable('a')
+            >>> bqm.linear
+            {'b': 2.}
+
+        """
+        linear = self.linear
+        if v in linear:
+            del linear[v]
+        else:
+            # nothing to remove
+            return
+
+        quadratic = self.quadratic
+        adj = self.adj
+
+        for u in adj[v]:
+            if (u, v) in quadratic:
+                del quadratic[(u, v)]
+            else:
+                del quadratic[(v, u)]
+
+            del adj[u][v]
+
+        del adj[v]
+
+        try:
+            # invalidates counterpart
+            del self._counterpart
```

Comment:
Why aren't we checking `hasattr` here (like line 520-523) instead of catching and passing on error in line 524-525?

Reply:
The following lines are only valid if _counterpart exists. We need to check because it can raise an attribute error on one but not on the other